src/v2/segment.v2.ejs

```diff
@@ -0,0 +1,159 @@
+
+// adaptation of:
+//   https://gist.github.com/sperand-io/4725e248a35d5005d68d810d8a8f7b29
+//   https://github.com/littledata/shopify-tracker/blob/ec4fac54cd56344980aa481800ee8dba550ea97d/src/customerPrivacyAPI/oneTrust.ts#L31
+
+const SEGMENT_WRITE_KEY = "<%%= sd.SEGMENT_WRITE_KEY %>";
+const SEGMENT_MAPPING = 'C0001'; // Segment itself assigned OneTrust Strictly Necessary category.
+let SEGMENT_DESTINATIONS = [];
+
+fetchDestinations(SEGMENT_WRITE_KEY).then(
+  destinations => {
+    SEGMENT_DESTINATIONS = destinations;
+
+    waitForOneTrust();
+
+    if (oneTrustReady()) {
+      getConsentAndLoadSegment();
+    }
+  }
+)
+
+// OneTrust calls OptanonWrapper when there's a consent change.
+window.OptanonWrapper = getConsentAndLoadSegment;
+
+function waitForOneTrust() {
+  let attempts = 0;
+  const maxAttempts = 100;
+
+  // wait 1sec at most.
+  if (!oneTrustReady()) {
+    const interval = setInterval(() => {
+      if (oneTrustReady() || attempts > maxAttempts) clearInterval(interval);
+      attempts++;
+    }, 10);
+  }
+}
+
+function oneTrustReady() {
+    // OneTrust is ready if OnetrustActiveGroups contains at least Strictly Necessary group.
+    if ((typeof window.OnetrustActiveGroups === 'string') && (window.OnetrustActiveGroups.split(',').includes('C0001'))) {
+      return true;
+    }
+    return false;
+}
+
+async function fetchDestinations(writeKey) {
+  let destinations = await fetchDestinationForWriteKey(writeKey);
+
+  destinations = [
+    ...destinations
+      .reduce((map, item) => {
+        if (item.id === "Repeater") return map; // remove Segment Repeater destinations
+        map.has(item["id"]) || map.set(item["id"], item);
+        return map;
+      }, new Map()) // return object
+      .values()
+  ];
+
+  return destinations;
+}
+
+function getConsentAndLoadSegment() {
+  let oneTrustConsent = '';
+
+  // OneTrust stores consent in OnetrustActiveGroups.
+  oneTrustConsent = window.OnetrustActiveGroups.split(',');
+
+  const destinationPreferences = setSegmentDestinationPref(oneTrustConsent, SEGMENT_DESTINATIONS);
+
+  conditionallyLoadAnalytics({
+    writeKey: SEGMENT_WRITE_KEY,
+    destinations: SEGMENT_DESTINATIONS,
+    destinationPreferences
+  });
+}
+
+function setSegmentDestinationPref(oneTrustConsent, destinations) {
+  // maps Segment destination's category to OneTrust cookie category.
+  const segmentToOneTrust= {
+    'SMS & Push Notifications': 'C0001', // OneTrust Strictly Necessary
```

Comment:
because Segment's category for Braze is `SMS & Push Notifications`

Reply:
Why map the destinations by category (rather than, say, name)? I could imagine some destinations with the same category having different degrees of necessity.

Reply:
Wanted to avoid coding Destination names in the script, but yeah, by category does seem coarse. Best if we can straight up tag each destination on Segment's UI with appropriate OneTrust category, and have Segment API return that. Time to involve @abhitip ! @abhitip - do you know if that's possible?

Reply:
Meanwhile, I will update the notion [doc](https://www.notion.so/artsy/Cookie-Consent-01dace52a765476ea4e9353260ce9a7f) with Segment destinations - a table that maps each destination to a OneTrust category.

Reply:
Finally getting to this - the table in your cookie consent doc looks good to me!

Reply:
Thanks @abhitip. I will update the code here to reflect the [table in the doc](https://www.notion.so/artsy/Cookie-Consent-01dace52a765476ea4e9353260ce9a7f#3ec509bfb40b4d08985f6d81a306f7bf). Let me open a ticket with Segment to see if we can add a OneTrust tag to the Destinations and have their API return that tag. That would be ideal.

Reply:
@abhitip -  Here's the reply from Segment. Apparently there's no way to tag Destinations. We will have to hardcode Destination-to-OneTrust-ID mapping in this script. This means we will have to remember to update this script for Destination changes.

```
Priscila (Segment)

Nov 16, 2021, 7:09 AM MST

Hi Jian,

Thanks for reaching out - happy to help here!

1) Is it possible to assign a custom attribute to each destination (via Segment UI), and have the API return that attribute along with the current ones (name/CreationName/category/...)? We want an attribute to store the OneTrust consent group (Necessary, Functional, Performance, etc.), so that the script can simply parse it from the API's response. The existing 'category' attribute doesn't work because there are multiple destinations of the same category that should be mapped differently. For example, we want the 'Amazon S3' destination (Analyticss) to map to 'Necessary', while other Analytics destinations to map to 'Performance'.


I'm afraid it's not currently possible to add custom attributes to destinations. I'm happy to pass on your feedback formally by submitting a feature request to our Engineering team. Let me know if you would like me to open a feature request? If you are able to share any specifics on the business need for this feature and the impact without this functionality, that would be helpful context for me to include in the request, too.

In the spirit of transparency, our teams are currently working through a backlog of feature requests right now. Our Product team does review every request, but I wouldn't be able to offer you a timeline as to if and when it would be looked at quite yet. Our Product team does review each and every feature and takes many things into consideration for planning which feature will be prioritized for the roadmap.


2) The API returns all our destinations except 'Redshift'. What is the reason?


By default, data warehouses doesn't populate in the integrations endpoint. Sorry about that. 


Priscila
Product Technical Lead @ Segment
```